### PR TITLE
[libSyntax] Restructure to more closely resemble the SwiftSyntax data structure

### PR DESCRIPTION
**Pull request by ahoppen**
After trying to use the libSyntax tree in the compiler pipeline, I uncovered several performance bottlenecks in the way libSyntax trees are represented. As a first step towards a more efficient representation of the syntax tree in C++, this PR refactors the current libSyntax tree to more closely resemble the way the tree is represented in SwiftSyntax, which has received more performance-related attention in the recent past.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
Discovered another clean-up commit that wanted to make it into this PR. Everything should be in now

@swift-ci Please test

**Comment by ahoppen** [comment]
Failure seems unrelated.

@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please ASAN test

**Comment by ahoppen** [comment]
ASAN test timed out (https://ci.swift.org/job/swift-PR-macos-ASAN-test/192/). Let’s start it again

@swift-ci Please ASAN test

**Comment by ahoppen** [comment]
ASAN build timed out again…

**Comment by ahoppen** [comment]
Giving the ASAN bot a third chance since it seems to be passing lately on other PRs but the point where it stalls doesn’t seem to be directly related to my changes. If it times out again, I’ll need to look into it.

@swift-ci Please ASAN test

**Review by rintaro — APPROVED** [review]
I haven't been able to look into the implementation line by line, but this looks good in general.

File: lib/Syntax/AbsoluteRawSyntax.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+//===--- AbsoluteRawSyntax.h ------------------------------------*- C++ -*-===//
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/Syntax/AbsoluteRawSyntax.cpp]: Whoops. Fixed it.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test linux platform

**Pull request merged**